### PR TITLE
MM-20787 : Destructive LHS leave channel menu option

### PR DESCRIPTION
**Pull request by AbhinavVihan**
#### Summary
Leave channel sidebar menu option appearing destructive

#### Ticket Link

  Fixes https://github.com/mattermost/mattermost-server/issues/20787

JIRA  -  https://mattermost.atlassian.net/browse/MM-45999

#### Related Pull Requests
None

#### Screenshots

Before --- 
Screenshot from 2022-08-31 21-20-39

After --- 
Screenshot from 2022-08-31 21-22-37

#### Release Note

```release-note
*Add destructive leave channel UI changes

```

### CODE REVIEW

File: components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx [hunk 1]
```diff
@@ -209,16 +209,32 @@ export class SidebarChannelMenu extends React.PureComponent<Props, State> {
...
+                        <Menu.ItemAction
+                            id={`leave-${channel.id}`}
+                            onClick={this.handleLeaveChannel}
+                            icon={<i className='icon-close'/>}
+                            text={leaveChannelText}
+                        />
+                    </Menu.Group>);
+            } else {
+                leaveChannel = (
+                    <Menu.Group>
```
Comment (michelengelen) [inline components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx:224]: we should be moving the `Menu.Group` wrapper to the return value as we do with the other items

File: components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx [hunk 2]
```diff
@@ -209,16 +209,32 @@ export class SidebarChannelMenu extends React.PureComponent<Props, State> {
...
+            if (channel.type === Constants.DM_CHANNEL || channel.type === Constants.GM_CHANNEL) {
+                leaveChannel =
+                    (<Menu.Group>
+                        <Menu.ItemAction
+                            id={`leave-${channel.id}`}
+                            onClick={this.handleLeaveChannel}
+                            icon={<i className='icon-close'/>}
+                            text={leaveChannelText}
+                        />
+                    </Menu.Group>);
```
Comment (michelengelen) [inline components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx:221]:
```suggestion
                leaveChannel = (
                    <Menu.ItemAction
                        id={`leave-${channel.id}`}
                        onClick={this.handleLeaveChannel}
                        icon={<i className='icon-close'/>}
                        text={leaveChannelText}
                    />
                );
```

File: components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx [hunk 3]
```diff
@@ -209,16 +209,32 @@ export class SidebarChannelMenu extends React.PureComponent<Props, State> {
...
-                    />
-                </Menu.Group>
-            );
+            if (channel.type === Constants.DM_CHANNEL || channel.type === Constants.GM_CHANNEL) {
+                leaveChannel =
+                    (<Menu.Group>
+                        <Menu.ItemAction
+                            id={`leave-${channel.id}`}
+                            onClick={this.handleLeaveChannel}
+                            icon={<i className='icon-close'/>}
```
Comment (michelengelen) [inline components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx:218]: we should be using the icon components from `@mattermost/compass-icons/components` here. It takes a `size` and `color` prop. `size` should be `16` and `color` should be `'currentColor'` so it does take the colro that is being set by the component for each state (default, hovering, active, etc.)

File: components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx [hunk 4]
```diff
@@ -209,16 +209,32 @@ export class SidebarChannelMenu extends React.PureComponent<Props, State> {
...
+                        <span
+                            onClick={this.handleLeaveChannel}
+                            id={`leave-${channel.id}`}
+                        >
+                            <span className='MenuItem__divider'/>
+                            <li className='destructive-leave-channel'>
+                                <i className='icon-close text-white'/>
+                                {leaveChannelText}
+                            </li>
+                        </span>
```
Comment (michelengelen) [inline components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx:234]: we can use the `Menu.ItemAction` component here as well instead. It has a prop called `isDangerous` that should add the correct semantic colors to it.

File: components/widgets/menu/menu_items/menu_item.scss [hunk 5]
```diff
@@ -354,3 +354,30 @@
...
+    }
+}
+
+.destructive-leave-channel:hover .text-white {
+    color: #fff !important;
+}
+
+.text-white {
+    color: var(--dnd-indicator) !important;
+}
```
Comment (michelengelen) [inline components/widgets/menu/menu_items/menu_item.scss:383]: when using the `Menu.ItemAction` with the `isDangerous` prop all of those styles are not needed

Comment (michelengelen) [reply components/widgets/menu/menu_items/menu_item.scss:383]: screenshot of the `Delete Category` item in the category dropdown menu in the LHS

Screenshot 2022-08-30 at 14 23 17

**Review by michelengelen — CHANGES_REQUESTED** [review]

**Comment by AbhinavVihan** [comment]
didn't know about the isDangerous props.
i will get back to you with the requested changes.

**Comment by AbhinavVihan** [comment]
Requestes changes are done accordingly.

File: components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx [hunk 6]
```diff
@@ -209,32 +209,16 @@ export class SidebarChannelMenu extends React.PureComponent<Props, State> {
...
-            if (channel.type === Constants.DM_CHANNEL || channel.type === Constants.GM_CHANNEL) {
-                leaveChannel =
+            leaveChannel =
                     (<Menu.Group>
                         <Menu.ItemAction
                             id={`leave-${channel.id}`}
                             onClick={this.handleLeaveChannel}
                             icon={<i className='icon-close'/>}
                             text={leaveChannelText}
+                            isDangerous={(channel.type === Constants.DM_CHANNEL || channel.type === Constants.GM_CHANNEL) === false}
```
Comment (michelengelen) [inline components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx:219]: no need to check for `=== false` ... you can just revert the boolean by using `!` like this:

```suggestion
                            isDangerous={!(channel.type === Constants.DM_CHANNEL || channel.type === Constants.GM_CHANNEL)}
```

File: components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx (hunk 3 above)
Comment (michelengelen) [reply components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx:217]: could you add this change in as well please?

File: components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx [hunk 7]
```diff
@@ -209,32 +209,16 @@ export class SidebarChannelMenu extends React.PureComponent<Props, State> {
 
         let leaveChannel;
         if (channel.name !== Constants.DEFAULT_CHANNEL) {
-            if (channel.type === Constants.DM_CHANNEL || channel.type === Constants.GM_CHANNEL) {
-                leaveChannel =
+            leaveChannel =
```
Comment (michelengelen) [inline components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx:212]: we typically leave the opening `(` in the same line with the assignment

```suggestion
            leaveChannel = (
```

I know that we have been inconsistent with this historically, so could you please add this change in as well? Thanks!

File: components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx [hunk 8]
```diff
@@ -209,32 +209,16 @@ export class SidebarChannelMenu extends React.PureComponent<Props, State> {
...
                     (<Menu.Group>
                         <Menu.ItemAction
                             id={`leave-${channel.id}`}
                             onClick={this.handleLeaveChannel}
                             icon={<i className='icon-close'/>}
                             text={leaveChannelText}
+                            isDangerous={(channel.type === Constants.DM_CHANNEL || channel.type === Constants.GM_CHANNEL) === false}
                         />
                     </Menu.Group>);
```
Comment (michelengelen) [inline components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx:221]: and then we close it with the same indentation as before:

```suggestion
                    </Menu.Group>
                );
```

**Review by michelengelen — CHANGES_REQUESTED** [review]
a few more little changes and we are good to go

File: components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx (hunk 7 above)
Comment (AbhinavVihan) [reply components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx]: No worries

**Comment by AbhinavVihan** [comment]
I've commited the requested changes.

File: components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx [hunk 9]
```diff
@@ -209,16 +209,17 @@ export class SidebarChannelMenu extends React.PureComponent<Props, State> {
...
-                            text={leaveChannelText}
-                            isDangerous={(channel.type === Constants.DM_CHANNEL || channel.type === Constants.GM_CHANNEL) === false}
-                        />
-                    </Menu.Group>);
+            leaveChannel = (
+                <Menu.Group>
+                    <Menu.ItemAction
+                        id={`leave-${channel.id}`}
+                        onClick={this.handleLeaveChannel}
+                        icon={<i className='icon-close'/>}
```
Comment (michelengelen) [inline components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx:217]: There is still the old implementation here with the `i` tag. we are deprecating these, so please use the icon component (`CloseIcon`) from `@mattermost/compass-icons/component`.

If you need help in implementing that fell free to reach out

**Review by michelengelen — CHANGES_REQUESTED** [review]

File: components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx (hunk 9 above)
Comment (AbhinavVihan) [reply components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx:217]: got it

Comment (AbhinavVihan) [reply components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx]: Migrated to ( CloseIcon ) from @mattermost/compass-icons/component.

File: components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx [hunk 10]
```diff
@@ -206,15 +210,30 @@ export class SidebarChannelMenu extends React.PureComponent<Props, State> {
...
                     <Menu.ItemAction
+                        onMouseEnter={() => this.setState({closeIconHovered: true})}
+                        onMouseLeave={() => this.setState({closeIconHovered: false})}
                         id={`leave-${channel.id}`}
                         onClick={this.handleLeaveChannel}
-                        icon={<i className='icon-close'/>}
+                        icon={
+                            <CloseIcon
+                                size={16}
+                                color={color()}
```
Comment (michelengelen) [inline components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx:234]: sry, but this adds a ton of complexity ... my fault... I just had a look at how the icon gets rendered inside the `MenuItem` component and it always wraps it in a `<span className='icon'>` component. We will postpone this once we run a campaign to replace all of the `i` tags with `svg` icon components.

If you could just revert your latest changes that will be fine for now.

Again I am sorry that I requested that change. I thought it is easier to implement, but apparently needs more work than anticipated.

File: components/widgets/menu/menu_items/menu_item_action.tsx [hunk 11]
```diff
@@ -8,6 +8,8 @@ import menuItem from './menu_item';
 
 type Props = {
     onClick: (e: React.MouseEvent) => void;
+    onMouseEnter?: (e: React.MouseEvent) => void;
```
Comment (michelengelen) [inline components/widgets/menu/menu_items/menu_item_action.tsx:11]: you can revert all of this too ... not needed anymore ... thanks!

**Review by michelengelen — CHANGES_REQUESTED** [review]
sry about that ... it really adds a whole lot of complexity

File: components/widgets/menu/menu_items/menu_item_action.tsx (hunk 11 above)
Comment (AbhinavVihan) [reply components/widgets/menu/menu_items/menu_item_action.tsx]: Done.

**Review by michelengelen — APPROVED** [review]
LGTM

**Review by AshishDhama — APPROVED** [review]

**Comment by AshishDhama** [comment]
@AbhinavVihan adding screenshots of the before and after would be great.

**Comment by AbhinavVihan** [comment]
ScreenShots were added in the description

**Review by jgilliam17 — APPROVED** [review]
Thank you @AbhinavVihan 
Tested, looks good to merge.
Verified destructive LHS leave channel menu option.
[E2E report](https://mattermost-cypress-report.s3.amazonaws.com/235812-a3ac796-onprem-ent-webapp-pr-11020-mm-ee-test:a3ac796/mochawesome.html) - no PR related failures.

**Pull request merged**